### PR TITLE
Update deny subnet without NSG & UDR policy

### PR DESCRIPTION
**Pull request by withstu**
## Overview/Summary

This PR fixes the issue that is described here: https://github.com/Azure/Enterprise-Scale/issues/407
Despite of the deny policies Deny-Subnet-Without-Udr & Deny-Subnet-Without-Nsg a vNet with a subnet without udr or nsg can be created in a single ARM deployment. After the deployment the subnets get be marked as non-compliant.

## This PR fixes/adds/changes/removes

1. https://github.com/Azure/Enterprise-Scale/issues/407

### From @withstu
After the fix I now getting a deny, when I try to create a vnet with a subnet without udr & nsg.

#### Azure Public
**Portal URL:** https://portal.azure.com/#blade/Microsoft_Azure_CreateUIDef/CustomDeploymentBlade/uri/https%3A%2F%2Fraw.githubusercontent.com%2Fwithstu%2FEnterprise-Scale%2Fmain%2FeslzArm%2FeslzArm.json/uiFormDefinitionUri/https%3A%2F%2Fraw.githubusercontent.com%2Fwithstu%2FEnterprise-Scale%2Fmain%2FeslzArm%2Feslz-portal.json

##### Evidence
image

image

image

image

image

image

image

image

#### Azure Fairfax (US Gov)
**Portal URL:** https://portal.azure.us/#blade/Microsoft_Azure_CreateUIDef/CustomDeploymentBlade/uri/https%3A%2F%2Fraw.githubusercontent.com%2Fwithstu%2FEnterprise-Scale%2Fmain%2FeslzArm%2FeslzArm.json/uiFormDefinitionUri/https%3A%2F%2Fraw.githubusercontent.com%2Fwithstu%2FEnterprise-Scale%2Fmain%2FeslzArm%2Ffairfaxeslz-portal.json

##### Evidence
See: https://github.com/Azure/Enterprise-Scale/pull/885#issuecomment-1011322492

#### Azure Mooncake (China)
**No portal experience**

##### Evidence
See: https://github.com/Azure/Enterprise-Scale/pull/885#issuecomment-1009609821

## As part of this Pull Request I have

- [x] Checked for duplicate [Pull Requests](https://github.com/Azure/Enterprise-Scale/pulls)
- [x] Associated it with relevant [issues](https://github.com/Azure/Enterprise-Scale/issues), for tracking and closure.
- [x] Ensured my code/branch is up-to-date with the latest changes in the `main` [branch](https://github.com/Azure/Enterprise-Scale/tree/main)
- [x] Performed testing and provided evidence.
- [x] Updated relevant and associated documentation.
- [x] Updated the ["What's New?"](https://github.com/Azure/Enterprise-Scale/wiki/Whats-new) wiki page (located: `/docs/wiki/whats-new.md`)

### CODE REVIEW

**Comment by jtracey93** [comment]
Thanks for raising this @withstu and suggesting a fix. Let me review and come back to you 👍

Also I saw you created this PR on the ESLZ Terraform module repo https://github.com/Azure/terraform-azurerm-caf-enterprise-scale/pull/249 which I will now close as the ESLZ Terraform module pulls down the policies from this repo as a source of truth 👍

But if approved here we can just update the policies downstream in the ESLZ Terraform module 👍

cc: @krowlandson & @matt-FFFFFF - for ESLZ Terraform module awareness

**Comment by jtracey93** [comment]
Hey @withstu as you've probably seen I've just updated your fork/branch with commit: https://github.com/withstu/Enterprise-Scale/commit/e0a822811ee598f4634b38927a1bbe09adaa9ef8

I tested this with the following bicep template and just commented out each resources to test all different combinations which now work correctly with the updated policy I put in the above commit 👍

```bicep
targetScope = 'resourceGroup'

param parRegion string = 'westeurope'

param parVnetWithNestedSubnetsName string = 'vnet-with-nested-subnets-2'

param parVnetBlankName string = 'vnet-blank2'

resource resVnetWithNestedSubnets 'Microsoft.Network/virtualNetworks@2021-05-01' = {
  name: parVnetWithNestedSubnetsName
  location: parRegion
  properties: {
    addressSpace: {
      addressPrefixes: [
        '10.0.0.0/16'
      ]
    }
    subnets: [
      {
        name: 'Subnet-1'
        properties: {
          addressPrefix: '10.0.0.0/24'
        }
      }
      {
        name: 'Subnet-2'
        properties: {
          addressPrefix: '10.0.1.0/24'
        }
      }
    ]
  }
}

// resource resBlankVnet 'Microsoft.Network/virtualNetworks@2021-05-01' = {
//   name: parVnetBlankName
//   location: parRegion
//   properties: {
//     addressSpace: {
//       addressPrefixes: [
//         '10.1.0.0/16'
//       ]
//     }
//   }
// }

// resource resSubnetForBlankVnet 'Microsoft.Network/virtualNetworks/subnets@2021-05-01' = {
//   parent: resBlankVnet
//   name: 'subnet-3'
//   properties: {
//     addressPrefix: '10.1.0.0/24'
//   }
// }

```

We also need to make the same changes to the other Azure environment deployments:

- [ ] Fairfax (US Gov)
- [ ] Mooncake (Azure China) 

And then test in all of them with a full deployment of the Azure Landing Zone Accelerator in each of them to ensure we haven't introduced any breaking changes.

I'm happy to own this and take it from here, but it may take a few days to get all the testing done etc.

Leave it with me 👍

And thanks again for raising and suggesting a fix 😎

**Comment by jtracey93** [comment]
Azure Public Testing Complete and can be seen here: https://github.com/Azure/Enterprise-Scale/pull/885#issue-815651157

**Comment by faister** [comment]
> Have updated all docs etc. required.
> 
> Now testing implementation across all 3 clouds.
> 
> * Azure Public = Me
> * Azure Mooncake = @faister
>               
>               Fai Lai
>               FTE
> * Azure Fairfax = @rspott
>               
>               Ryan Spott
>               FTE
> 
> Will report back when testing complete and then PR can be reviewed and approved 👍

@jtracey93 Completed testing in Azure China

Considering that ALZ deployment in Azure China is performed by using DIY ARM deployment, I am attaching the deployment provisioning output here.  
[EslzMC-policy-pr885-DeploymentSucceeded.txt](https://github.com/Azure/Enterprise-Scale/files/7843804/EslzMC-policy-pr885-DeploymentSucceeded.txt)

image

image

Then did both policies assignment onto my ALZ management group scope. Tried to create a vnet, review resulted in a failed validation which is expected. The vnet resource was disallowed by both policies.

image

File: eslzArm/managementGroupTemplates/policyDefinitions/china/mcPolicies.json [hunk 1]
```diff
@@ -6920,7 +6920,7 @@
         },
         {
           "properties": {
-            "description": "This policy denies the creation of a subsnet with out an Network Security Group. NSG help to protect traffic across subnet-level.",
+            "description": "This policy denies the creation of a subnet with out an Network Security Group. NSG help to protect traffic across subnet-level.",
```
Comment (faister) [inline eslzArm/managementGroupTemplates/policyDefinitions/china/mcPolicies.json:6923]: "This policy denies the creation of a subnet without a Network Security Group (NSG). NSG helps to protect traffic across subnet-level."

File: eslzArm/managementGroupTemplates/policyDefinitions/policies.json [hunk 2]
```diff
@@ -7048,7 +7048,7 @@
                 },
                 {
                     "properties": {
-                        "description": "This policy denies the creation of a subsnet with out an Network Security Group. NSG help to protect traffic across subnet-level.",
+                        "description": "This policy denies the creation of a subnet with out an Network Security Group. NSG help to protect traffic across subnet-level.",
```
Comment (faister) [inline eslzArm/managementGroupTemplates/policyDefinitions/policies.json:7051]: "This policy denies the creation of a subnet without a Network Security Group (NSG). NSG helps to protect traffic across subnet-level."

File: eslzArm/managementGroupTemplates/policyDefinitions/china/mcPolicies.json [hunk 3]
```diff
@@ -6985,7 +7015,7 @@
             "mode": "All",
             "description": "This policy denies the creation of a subnet with out a User Defined Route.",
```
Comment (faister) [inline eslzArm/managementGroupTemplates/policyDefinitions/china/mcPolicies.json:7016]: "This policy denies the creation of a subnet without a User Defined Route (UDR)."

File: eslzArm/managementGroupTemplates/policyDefinitions/policies.json [hunk 4]
```diff
@@ -7113,7 +7143,7 @@
                         "mode": "All",
                         "description": "This policy denies the creation of a subnet with out a User Defined Route.",
```
Comment (faister) [inline eslzArm/managementGroupTemplates/policyDefinitions/policies.json:7144]: "This policy denies the creation of a subnet without a User Defined Route (UDR)."

**Review by faister — CHANGES_REQUESTED** [review]
suggesting some changes on the description of the policies  `Deny-Subnet-Without-Nsg` & `Deny-Subnet-Without-Udr`

File: eslzArm/managementGroupTemplates/policyDefinitions/china/mcPolicies.json (hunk 1 above)
Comment (jtracey93) [reply eslzArm/managementGroupTemplates/policyDefinitions/china/mcPolicies.json:6923]: This is already as per your suggestion in the latest commit in this branch: https://github.com/withstu/Enterprise-Scale/blob/da051b551a26e28a2ee248e64a8e73e4a8e154f0/eslzArm/managementGroupTemplates/policyDefinitions/china/mcPolicies.json#L6923

File: eslzArm/managementGroupTemplates/policyDefinitions/china/mcPolicies.json (hunk 3 above)
Comment (jtracey93) [reply eslzArm/managementGroupTemplates/policyDefinitions/china/mcPolicies.json]: updated

**Comment by rspott** [comment]
Completed testing in Fairfax.

Subnets should have a User Defined Route:

```
{
  "properties": {
    "displayName": "Subnets should have a User Defined Route",
    "policyType": "Custom",
    "mode": "All",
    "description": "This policy denies the creation of a subnet with out a User Defined Route.",
    "metadata": {
      "version": "1.1.0",
      "category": "Network",
      "createdBy": "0b337910-de41-46f7-8a3c-83c8d78f039a",
      "createdOn": "2022-01-10T22:29:52.8384786Z",
      "updatedBy": "0b337910-de41-46f7-8a3c-83c8d78f039a",
      "updatedOn": "2022-01-10T23:07:27.155697Z"
    },
    "parameters": {
      "effect": {
        "type": "String",
        "metadata": {
          "displayName": "Effect",
          "description": "Enable or disable the execution of the policy"
        },
        "allowedValues": [
          "Audit",
          "Deny",
          "Disabled"
        ],
        "defaultValue": "Deny"
      },
      "excludedSubnets": {
        "type": "Array",
        "metadata": {
          "displayName": "Excluded Subnets",
          "description": "Array of subnet names that are excluded from this policy"
        },
        "defaultValue": [
          "AzureBastionSubnet"
        ]
      }
    },
    "policyRule": {
      "if": {
        "anyOf": [
          {
            "allOf": [
              {
                "equals": "Microsoft.Network/virtualNetworks",
                "field": "type"
              },
              {
                "count": {
                  "field": "Microsoft.Network/virtualNetworks/subnets[*]",
                  "where": {
                    "allOf": [
                      {
                        "exists": "false",
                        "field": "Microsoft.Network/virtualNetworks/subnets[*].routeTable.id"
                      },
                      {
                        "field": "Microsoft.Network/virtualNetworks/subnets[*].name",
                        "notIn": "[parameters('excludedSubnets')]"
                      }
                    ]
                  }
                },
                "notEquals": 0
              }
            ]
          },
          {
            "allOf": [
              {
                "field": "type",
                "equals": "Microsoft.Network/virtualNetworks/subnets"
              },
              {
                "field": "name",
                "notIn": "[parameters('excludedSubnets')]"
              },
              {
                "field": "Microsoft.Network/virtualNetworks/subnets/routeTable.id",
                "exists": "false"
              }
            ]
          }
        ]
      },
      "then": {
        "effect": "[parameters('effect')]"
      }
    }
  },
  "id": "/providers/Microsoft.Management/managementGroups/test30/providers/Microsoft.Authorization/policyDefinitions/Deny-Subnet-Without-Udr",
  "type": "Microsoft.Authorization/policyDefinitions",
  "name": "Deny-Subnet-Without-Udr",
  "systemData": {
    "createdBy": "drspott@poc.cmmcmission365.com",
    "createdByType": "User",
    "createdAt": "2022-01-10T22:29:52.7089658Z",
    "lastModifiedBy": "drspott@poc.cmmcmission365.com",
    "lastModifiedByType": "User",
    "lastModifiedAt": "2022-01-10T23:07:26.9373516Z"
  }
}
```

And 

Subnets should have a Network Security Group:

```
{
  "properties": {
    "displayName": "Subnets should have a Network Security Group",
    "policyType": "Custom",
    "mode": "All",
    "description": "This policy denies the creation of a subsnet with out an Network Security Group. NSG help to protect traffic across subnet-level.",
    "metadata": {
      "version": "2.0.0",
      "category": "Network",
      "createdBy": "0b337910-de41-46f7-8a3c-83c8d78f039a",
      "createdOn": "2022-01-10T22:29:52.7435233Z",
      "updatedBy": "0b337910-de41-46f7-8a3c-83c8d78f039a",
      "updatedOn": "2022-01-10T23:07:29.0526971Z"
    },
    "parameters": {
      "effect": {
        "type": "String",
        "metadata": {
          "displayName": "Effect",
          "description": "Enable or disable the execution of the policy"
        },
        "allowedValues": [
          "Audit",
          "Deny",
          "Disabled"
        ],
        "defaultValue": "Deny"
      },
      "excludedSubnets": {
        "type": "Array",
        "metadata": {
          "displayName": "Excluded Subnets",
          "description": "Array of subnet names that are excluded from this policy"
        },
        "defaultValue": [
          "GatewaySubnet",
          "AzureFirewallSubnet",
          "AzureFirewallManagementSubnet"
        ]
      }
    },
    "policyRule": {
      "if": {
        "anyOf": [
          {
            "allOf": [
              {
                "equals": "Microsoft.Network/virtualNetworks",
                "field": "type"
              },
              {
                "count": {
                  "field": "Microsoft.Network/virtualNetworks/subnets[*]",
                  "where": {
                    "allOf": [
                      {
                        "exists": "false",
                        "field": "Microsoft.Network/virtualNetworks/subnets[*].networkSecurityGroup.id"
                      },
                      {
                        "field": "Microsoft.Network/virtualNetworks/subnets[*].name",
                        "notIn": "[parameters('excludedSubnets')]"
                      }
                    ]
                  }
                },
                "notEquals": 0
              }
            ]
          },
          {
            "allOf": [
              {
                "field": "type",
                "equals": "Microsoft.Network/virtualNetworks/subnets"
              },
              {
                "field": "name",
                "notIn": "[parameters('excludedSubnets')]"
              },
              {
                "field": "Microsoft.Network/virtualNetworks/subnets/networkSecurityGroup.id",
                "exists": "false"
              }
            ]
          }
        ]
      },
      "then": {
        "effect": "[parameters('effect')]"
      }
    }
  },
  "id": "/providers/Microsoft.Management/managementGroups/test30/providers/Microsoft.Authorization/policyDefinitions/Deny-Subnet-Without-Nsg",
  "type": "Microsoft.Authorization/policyDefinitions",
  "name": "Deny-Subnet-Without-Nsg",
  "systemData": {
    "createdBy": "drspott@poc.cmmcmission365.com",
    "createdByType": "User",
    "createdAt": "2022-01-10T22:29:52.709722Z",
    "lastModifiedBy": "drspott@poc.cmmcmission365.com",
    "lastModifiedByType": "User",
    "lastModifiedAt": "2022-01-10T23:07:28.9888333Z"
  }
}
```

**Comment by jtracey93** [comment]
@Azure/enterprisescale-vteam & @Azure/customer-architecture-team this is now ready for review and merging, if approved 👍

**Review by matt-FFFFFF — APPROVED** [review]

**Pull request merged**